### ISSUE TITLE
Automatic structural conversion can break type-safety at generic call-sites

### ISSUE THREAD

**Issue by ZachBray** [Microsoft/TypeScript#6389]
Automatic structural conversion causes different type-checking results at the call-sites of generic functions. 

In this case, the generic constraints are not taken into account:

```
interface A<T1, T2> {}
interface B<T1, T2> {}

function foo<T1, T2, T3>(x:T1, f:A<T1, T2>) {}

var addOne : A<string, string> = {}
var addOne2 : B<string, string> = {}

foo(1, addOne) // <-- ERROR HERE
foo(1, addOne2) // <-- FINE HERE (BUT SHOULD SHOW ERROR TOO?)
```

I would expect both of the bottom lines to have errors.

**Comment by ZachBray** [comment]
BTW this was discovered in http://www.typescriptlang.org/Playground
I assume v1.7

**Comment by RyanCavanaugh** [comment]
See #5451 or any other issue referencing the fact that empty types are interchangeable.

If you added a member to `A` / `B` that consumed `T1` or `T2` you'd see an error.
